File: Lab 2/C++/class.cpp
#include "ClassLib.h"
#include <cmath>

//Конструктор
LineSegment::LineSegment()
{
    // Виділяємо пам'ять для координат
    x1 = new double(0); 
    y1 = new double(0);
    x2 = new double(0);
    y2 = new double(0);
}

//Конструктор із параметрами
LineSegment::LineSegment(double x1, double y1, double x2, double y2)
{
    //Виділяємо пам'ять для координат
    this->x1 = new double(x1);  
    this->y1 = new double(y1);
    this->x2 = new double(x2);
    this->y2 = new double(y2);
}

//Конструктор копіювання, глибоке копіювання
LineSegment::LineSegment(const LineSegment& other)
{
    //Копіюємо значення і виділяємо пам'ять для нових об'єктів
    x1 = new double(*other.x1);
    y1 = new double(*other.y1);
    x2 = new double(*other.x2);
    y2 = new double(*other.y2);
}

//Конструктор перенесення
LineSegment::LineSegment(LineSegment&& other) noexcept
{
    //Переміщення вказівників
    x1 = other.x1;
    y1 = other.y1;
    x2 = other.x2;
    y2 = other.y2;

    //Обнуляємо вказівники в об'єкті, з якого відбувається переміщення
    other.x1 = nullptr;
    other.y1 = nullptr;
    other.x2 = nullptr;
    other.y2 = nullptr;
}

//Деструктор
LineSegment::~LineSegment() 
{
    //Звільняємо пам'ять, якщо вона була виділена
    delete x1;
    delete y1;
    delete x2;
    delete y2;

}

//Обчислення довжини відрізка
double LineSegment::length() const {
    return sqrt(pow(*x2 - *x1, 2) + pow(*y2 - *y1, 2));
}

// Метод для обчислення кута з віссю OX (в градусах)
double LineSegment::angleOX() const {
    return atan2(*y2 - *y1, *x2 - *x1) * 180.0 / 3.14;
}

File: Lab 2/C#/class.cs
using System;

public class LineSegment
{
    private double x1;
    private double y1;
    private double x2;
    private double y2;

    //Конструктор за замовчуванням
    public LineSegment()
    {
        x1 = x2 = y1 = y2 = 0;
    }

    //Конструктор із параметрами 
    public LineSegment(double x1, double y1, double x2, double y2)
    {
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
    }

    //Конструктор копіювання
    public LineSegment(LineSegment other)
    {
        x1 = other.x1;
        y1 = other.y1;
        x2 = other.x2;
        y2 = other.y2;
    }

    //Конструктор перенесення 
    public LineSegment(ref LineSegment other)
    {
        x1 = other.x1;
        y1 = other.y1;
        x2 = other.x2;
        y2 = other.y2;

        // Обнуляємо значення в іншому об'єкті
        other.x1 = 0;
        other.y1 = 0;
        other.x2 = 0;
        other.y2 = 0;
    }

    //Деструктор 
    ~LineSegment()
    {

    }

    public double Length()
    {
        return Math.Sqrt(Math.Pow(x2 - x1, 2) + Math.Pow(y2 - y1, 2));
    }

    public double AngleOX()
    {
        return Math.Atan2(y2 - y1, x2 - x1) * 180.0 / Math.PI;
    }
}
